File: Indexed Search/Basic Indexed Search/Sparse Index Search/SparseIndexSearch.java
import java.util.*;

public class SparseIndexSearch {

static class Index {

int key;

int pos;

Index(
int k,
int p
){

key=k;

pos=p;

}

}

static int search(
int[] arr,
int target
){

int n=
arr.length;

int step=
(int)
Math.sqrt(
n
);

List<Index>
idx=
new ArrayList<>();

for(
int i=0;
i<n;
i+=step
){

idx.add(
new Index(
arr[i],
i
)
);

}

int start=0;

for(
Index x:
idx
){

if(
x.key
>
target
)
break;

start=
x.pos;

}

int end=
Math.min(
start
+
step
-
1,
n-1
);

for(
int i=start;
i<=end;
i++
){

if(
arr[i]
==
target
)
return i;

}

return -1;

}

public static void main(
String[] args
){

int[] arr=
{
2,5,8,12,
16,20,24,
28,32,40
};

System.out.println(
search(
arr,
24
)
);

}

}

File: Indexed Search/Basic Indexed Search/Sparse Index Search/SparseIndexSearch.js
function sparseIndexSearch(
arr,
target
){

let n=
arr.length;

let step=
Math.floor(
Math.sqrt(
n
)
);

let index=[];

for(
let i=0;
i<n;
i+=step
){

index.push(
{
key:
arr[i],
pos:i
}
);

}

let start=0;

for(
let x
of
index
){

if(
x.key
>
target
)
break;

start=
x.pos;

}

let end=
Math.min(
start
+
step
-
1,
n-1
);

for(
let i=
start;
i<=end;
i++
){

if(
arr[i]
===
target
)
return i;

}

return -1;

}

const arr=
[
2,5,8,12,
16,20,24,
28,32,40
];

console.log(
sparseIndexSearch(
arr,
24
)
);
